File: src/chatty/knowledge.py
"""Chatty 知识库：FTS5 trigram + 单字索引 + JSONL 全量导入 + 三层回退检索。

规格：specs/stores.md §5。构造函数接收已打开的连接（生产复用 commerce.database）；
importJsonl 走 BEGIN IMMEDIATE 全量重建；search 为 catch-all 口径（TS 权威，
decisions.md §6：任何异常 → knowledge_search_unavailable）。
"""

from __future__ import annotations

import json
import re
import sqlite3
import threading
from pathlib import Path

from pydantic import ValidationError

from chatty.contracts import KnowledgeRecord, KnowledgeSearchResult
from chatty.store import database_write_lock, write_transaction

_SEARCHABLE_COLUMNS = ("title", "summary", "body", "tags")


class KnowledgeStore:
    """Imports seller-authored chunks and exposes one bounded lexical search seam."""

    def __init__(self, connection: sqlite3.Connection) -> None:
        self.database = connection
        connection.row_factory = sqlite3.Row
        main_file = self._main_database_file(connection)
        self._write_lock = database_write_lock(main_file) if main_file else threading.RLock()
        connection.execute(
            """
            CREATE VIRTUAL TABLE IF NOT EXISTS knowledge_fts USING fts5(
                id UNINDEXED,
                title,
                summary,
                body,
                source UNINDEXED,
                tags,
                tokenize='trigram'
            )
            """
        )
        connection.execute(
            """
            CREATE TABLE IF NOT EXISTS knowledge_character_index (
                record_rowid INTEGER NOT NULL,
                character TEXT NOT NULL,
                PRIMARY KEY (record_rowid, character)
            )
            """
        )
        connection.execute(
            """
            CREATE INDEX IF NOT EXISTS knowledge_character_lookup
            ON knowledge_character_index (character, record_rowid)
            """
        )

    def import_jsonl(self, source_path: str | Path) -> int:
        """全量重建：解析阶段任何失败都不动库；重建阶段在 BEGIN IMMEDIATE 事务内。"""
        records: list[KnowledgeRecord] = []
        seen_ids: set[str] = set()
        for line_number, line in enumerate(
            Path(source_path).read_text(encoding="utf-8").splitlines(), start=1
        ):
            if not line.strip():
                continue
            try:
                record = KnowledgeRecord.model_validate_json(line)
            except ValidationError as error:
                raise ValueError(f"invalid knowledge record on line {line_number}") from error
            if record.id in seen_ids:
                raise ValueError(f"duplicate knowledge id on line {line_number}: {record.id}")
            seen_ids.add(record.id)
            records.append(record)

        with write_transaction(self.database, self._write_lock) as connection:
            connection.execute("DELETE FROM knowledge_fts")
            connection.execute("DELETE FROM knowledge_character_index")
            for rowid, record in enumerate(records, start=1):
                tags_json = json.dumps(record.tags, ensure_ascii=False, separators=(",", ":"))
                connection.execute(
                    """
                    INSERT INTO knowledge_fts (rowid, id, title, summary, body, source, tags)
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        rowid,
                        record.id,
                        record.title,
                        record.summary,
                        record.body,
                        record.source,
                        tags_json,
                    ),
                )
                connection.executemany(
                    """
                    INSERT INTO knowledge_character_index (record_rowid, character)
                    VALUES (?, ?)
                    """,
                    [
                        (rowid, character)
                        for character in self._record_characters(record, tags_json)
                    ],
                )
        return len(records)

    def search(self, query: str, *, limit: int) -> KnowledgeSearchResult:
        normalized = " ".join(query.split())
        if not normalized:
            return KnowledgeSearchResult(status="ok", query=query, results=[])
        if len(normalized) > 500:
            return KnowledgeSearchResult(
                status="error",
                query=query,
                results=[],
                error="invalid_knowledge_query",
            )
        safe_limit = max(1, min(limit, 10))
        try:
            rows = self._search_rows(normalized, safe_limit)
            return KnowledgeSearchResult(
                status="ok",
                query=query,
                results=[self._record_from_row(row) for row in rows],
            )
        except Exception:
            # TS catch-all：SQL 错误、损坏行、契约不符统一降级，不外抛。
            return KnowledgeSearchResult(
                status="error",
                query=query,
                results=[],
                error="knowledge_search_unavailable",
            )

    # ------------------------------------------------------------------
    # 三层检索
    # ------------------------------------------------------------------

    def _search_rows(self, query: str, limit: int) -> list[sqlite3.Row]:
        fragments = self._lexical_fragments(query)
        if fragments:
            match_expression = " OR ".join(
                '"{}"'.format(fragment.replace('"', '""')) for fragment in fragments
            )
            rows = self.database.execute(
                """
                SELECT id, title, summary, body, source, tags
                FROM knowledge_fts
                WHERE knowledge_fts MATCH ?
                ORDER BY bm25(knowledge_fts), rowid
                LIMIT ?
                """,
                (match_expression, limit),
            ).fetchall()
            if rows:
                return rows
        terms = self._query_terms(query)
        exact_rows = self._short_term_rows(terms, limit)
        if exact_rows:
            return exact_rows
        return self._fuzzy_rows(terms, limit)

    @staticmethod
    def _lexical_fragments(query: str) -> list[str]:
        fragments: list[str] = []
        for term in re.findall(r"\w+", query):
            if len(term) < 3:
                continue
            fragments.extend(term[index : index + 3] for index in range(len(term) - 2))
        return list(dict.fromkeys(fragments))

    @staticmethod
    def _query_terms(query: str) -> list[str]:
        return re.findall(r"\w+", query)

    def _short_term_rows(self, terms: list[str], limit: int) -> list[sqlite3.Row]:
        if not terms:
            return []
        term_clauses: list[str] = []
        score_clauses: list[str] = []
        where_params: list[str] = []
        score_params: list[str] = []
        for term in terms:
            escaped = term.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
            pattern = f"%{escaped}%"
            clause = " OR ".join(f"{column} LIKE ? ESCAPE '\\'" for column in _SEARCHABLE_COLUMNS)
            term_clauses.append(f"({clause})")
            where_params.extend([pattern] * len(_SEARCHABLE_COLUMNS))
            score_clauses.append(f"CASE WHEN {clause} THEN 1 ELSE 0 END")
            score_params.extend([pattern] * len(_SEARCHABLE_COLUMNS))
        statement = f"""
            SELECT id, title, summary, body, source, tags,
                   ({" + ".join(score_clauses)}) AS match_count
            FROM knowledge_fts
            WHERE {" OR ".join(term_clauses)}
            ORDER BY match_count DESC, rowid
            LIMIT ?
        """
        return self.database.execute(statement, (*score_params, *where_params, limit)).fetchall()

    def _fuzzy_rows(self, terms: list[str], limit: int) -> list[sqlite3.Row]:
        fuzzy_terms = [term for term in terms if 2 <= len(term) <= 10]
        characters = list(dict.fromkeys("".join(fuzzy_terms)))
        if not characters:
            return []
        placeholders = ", ".join("?" for _ in characters)
        candidates = self.database.execute(
            f"""
            SELECT knowledge_fts.id, title, summary, body, source, tags, knowledge_fts.rowid
            FROM knowledge_fts
            JOIN (
                SELECT record_rowid, COUNT(*) AS shared_characters
                FROM knowledge_character_index
                WHERE character IN ({placeholders})
                GROUP BY record_rowid
                ORDER BY shared_characters DESC, record_rowid
                LIMIT 20
            ) AS candidate ON candidate.record_rowid = knowledge_fts.rowid
            """,
            characters,
        ).fetchall()
        ranked: list[tuple[int, int, sqlite3.Row]] = []
        for row in candidates:
            searchable = f"{row['title']} {row['summary']} {row['tags']}"
            score = sum(1 for term in fuzzy_terms if self._has_close_substring(searchable, term))
            if score:
                ranked.append((-score, int(row["rowid"]), row))
        ranked.sort(key=lambda item: (item[0], item[1]))
        return [row for _, _, row in ranked[:limit]]

    # ------------------------------------------------------------------
    # 帮助函数
    # ------------------------------------------------------------------

    @staticmethod
    def _record_characters(record: KnowledgeRecord, tags_json: str) -> list[str]:
        # 注意：不含 body、不含 source；tags 用与存储相同的 JSON 串。
        searchable = f"{record.title} {record.summary} {tags_json}"
        return sorted(set(re.findall(r"\w", searchable)))

    @staticmethod
    def _has_close_substring(value: str, term: str) -> bool:
        for width in {max(1, len(term) - 1), len(term), len(term) + 1}:
            for start in range(max(0, len(value) - width + 1)):
                if KnowledgeStore._edit_distance(value[start : start + width], term) <= 1:
                    return True
        return False

    @staticmethod
    def _edit_distance(left: str, right: str) -> int:
        previous = list(range(len(right) + 1))
        for left_index, left_character in enumerate(left, start=1):
            current = [left_index]
            for right_index, right_character in enumerate(right, start=1):
                current.append(
                    min(
                        current[-1] + 1,
                        previous[right_index] + 1,
                        previous[right_index - 1] + (left_character != right_character),
                    )
                )
            previous = current
        return previous[-1]

    @staticmethod
    def _record_from_row(row: sqlite3.Row) -> KnowledgeRecord:
        return KnowledgeRecord(
            id=row["id"],
            title=row["title"],
            summary=row["summary"],
            body=row["body"],
            source=row["source"],
            tags=json.loads(row["tags"]),
        )

    @staticmethod
    def _main_database_file(connection: sqlite3.Connection) -> str:
        for row in connection.execute("PRAGMA database_list"):
            if row["name"] == "main":
                return str(row["file"] or "")
        return ""

File: apps/api/src/knowledge.ts
import { readFileSync } from "node:fs";
import { DatabaseSync } from "node:sqlite";
import {
  KnowledgeRecordSchema,
  KnowledgeSearchResultSchema,
  type KnowledgeRecord,
  type KnowledgeSearchResult,
} from "@chatty/contracts";
import { type SqliteRow as Row, text } from "./sqlite-row.js";

function recordFromRow(row: Row): KnowledgeRecord {
  return KnowledgeRecordSchema.parse({
    id: text(row, "id"),
    title: text(row, "title"),
    summary: text(row, "summary"),
    body: text(row, "body"),
    source: text(row, "source"),
    tags: JSON.parse(text(row, "tags")) as unknown,
  });
}

function lexicalFragments(query: string): string[] {
  const fragments: string[] = [];
  for (const term of query.match(/[\p{L}\p{N}_]+/gu) ?? []) {
    if (term.length < 3) continue;
    for (let index = 0; index <= term.length - 3; index += 1) {
      const fragment = term.slice(index, index + 3);
      if (!fragments.includes(fragment)) fragments.push(fragment);
    }
  }
  return fragments;
}

function queryTerms(query: string): string[] {
  return query.match(/[\p{L}\p{N}_]+/gu) ?? [];
}

function editDistance(left: string, right: string): number {
  let previous = Array.from({ length: right.length + 1 }, (_, index) => index);
  for (let leftIndex = 1; leftIndex <= left.length; leftIndex += 1) {
    const current = [leftIndex];
    for (let rightIndex = 1; rightIndex <= right.length; rightIndex += 1) {
      current.push(
        Math.min(
          current[current.length - 1]! + 1,
          previous[rightIndex]! + 1,
          previous[rightIndex - 1]! +
            Number(left[leftIndex - 1] !== right[rightIndex - 1]),
        ),
      );
    }
    previous = current;
  }
  return previous[previous.length - 1]!;
}

function hasCloseSubstring(value: string, term: string): boolean {
  const widths = new Set([
    Math.max(1, term.length - 1),
    term.length,
    term.length + 1,
  ]);
  for (const width of widths) {
    for (
      let start = 0;
      start < Math.max(0, value.length - width + 1);
      start += 1
    ) {
      if (editDistance(value.slice(start, start + width), term) <= 1)
        return true;
    }
  }
  return false;
}

export class KnowledgeStore {
  constructor(private readonly database: DatabaseSync) {
    this.database.exec(`
      CREATE VIRTUAL TABLE IF NOT EXISTS knowledge_fts USING fts5(
          id UNINDEXED,
          title,
          summary,
          body,
          source UNINDEXED,
          tags,
          tokenize='trigram'
      );
      CREATE TABLE IF NOT EXISTS knowledge_character_index (
          record_rowid INTEGER NOT NULL,
          character TEXT NOT NULL,
          PRIMARY KEY (record_rowid, character)
      );
      CREATE INDEX IF NOT EXISTS knowledge_character_lookup
          ON knowledge_character_index (character, record_rowid);
    `);
  }

  importJsonl(sourcePath: string): number {
    const records: KnowledgeRecord[] = [];
    const seenIds = new Set<string>();
    const lines = readFileSync(sourcePath, "utf8").split(/\r?\n/);
    for (const [index, line] of lines.entries()) {
      if (!line.trim()) continue;
      let record: KnowledgeRecord;
      try {
        record = KnowledgeRecordSchema.parse(JSON.parse(line) as unknown);
      } catch {
        throw new Error(`invalid knowledge record on line ${index + 1}`);
      }
      if (seenIds.has(record.id)) {
        throw new Error(
          `duplicate knowledge id on line ${index + 1}: ${record.id}`,
        );
      }
      seenIds.add(record.id);
      records.push(record);
    }

    this.database.exec("BEGIN IMMEDIATE");
    try {
      this.database.exec(
        "DELETE FROM knowledge_fts; DELETE FROM knowledge_character_index;",
      );
      const insertRecord = this.database.prepare(
        `INSERT INTO knowledge_fts (rowid, id, title, summary, body, source, tags)
         VALUES (?, ?, ?, ?, ?, ?, ?)`,
      );
      const insertCharacter = this.database.prepare(
        `INSERT INTO knowledge_character_index (record_rowid, character)
         VALUES (?, ?)`,
      );
      for (const [index, record] of records.entries()) {
        const rowId = index + 1;
        insertRecord.run(
          rowId,
          record.id,
          record.title,
          record.summary,
          record.body,
          record.source,
          JSON.stringify(record.tags),
        );
        const searchable = `${record.title} ${record.summary} ${JSON.stringify(record.tags)}`;
        const characters = [
          ...new Set(searchable.match(/[\p{L}\p{N}_]/gu) ?? []),
        ].sort();
        for (const character of characters)
          insertCharacter.run(rowId, character);
      }
      this.database.exec("COMMIT");
      return records.length;
    } catch (error) {
      this.database.exec("ROLLBACK");
      throw error;
    }
  }

  search(query: string, limit: number): KnowledgeSearchResult {
    const normalized = query.trim().replace(/\s+/g, " ");
    if (!normalized) return { status: "ok", query, results: [], error: null };
    if (normalized.length > 500) {
      return {
        status: "error",
        query,
        results: [],
        error: "invalid_knowledge_query",
      };
    }
    try {
      return KnowledgeSearchResultSchema.parse({
        status: "ok",
        query,
        results: this.searchRows(
          normalized,
          Math.max(1, Math.min(limit, 10)),
        ).map(recordFromRow),
        error: null,
      });
    } catch {
      return {
        status: "error",
        query,
        results: [],
        error: "knowledge_search_unavailable",
      };
    }
  }

  private searchRows(query: string, limit: number): Row[] {
    const fragments = lexicalFragments(query);
    if (fragments.length > 0) {
      const expression = fragments
        .map((fragment) => `"${fragment.replaceAll('"', '""')}"`)
        .join(" OR ");
      const rows = this.database
        .prepare(
          `SELECT id, title, summary, body, source, tags
           FROM knowledge_fts
           WHERE knowledge_fts MATCH ?
           ORDER BY bm25(knowledge_fts), rowid
           LIMIT ?`,
        )
        .all(expression, limit) as Row[];
      if (rows.length > 0) return rows;
    }
    const terms = queryTerms(query);
    const exact = this.shortTermRows(terms, limit);
    return exact.length > 0 ? exact : this.fuzzyRows(terms, limit);
  }

  private shortTermRows(terms: string[], limit: number): Row[] {
    if (terms.length === 0) return [];
    const columns = ["title", "summary", "body", "tags"];
    const termClauses: string[] = [];
    const scoreClauses: string[] = [];
    const params: string[] = [];
    const scoreParams: string[] = [];
    for (const term of terms) {
      const escaped = term
        .replaceAll("\\", "\\\\")
        .replaceAll("%", "\\%")
        .replaceAll("_", "\\_");
      const pattern = `%${escaped}%`;
      const clause = columns
        .map((column) => `${column} LIKE ? ESCAPE '\\'`)
        .join(" OR ");
      termClauses.push(`(${clause})`);
      scoreClauses.push(`CASE WHEN ${clause} THEN 1 ELSE 0 END`);
      params.push(...columns.map(() => pattern));
      scoreParams.push(...columns.map(() => pattern));
    }
    return this.database
      .prepare(
        `SELECT id, title, summary, body, source, tags,
                (${scoreClauses.join(" + ")}) AS match_count
         FROM knowledge_fts
         WHERE ${termClauses.join(" OR ")}
         ORDER BY match_count DESC, rowid
         LIMIT ?`,
      )
      .all(...scoreParams, ...params, limit) as Row[];
  }

  private fuzzyRows(terms: string[], limit: number): Row[] {
    const fuzzyTerms = terms.filter(
      (term) => term.length >= 2 && term.length <= 10,
    );
    const characters = [...new Set(fuzzyTerms.join(""))];
    if (characters.length === 0) return [];
    const candidates = this.database
      .prepare(
        `SELECT knowledge_fts.id, title, summary, body, source, tags, knowledge_fts.rowid
         FROM knowledge_fts
         JOIN (
           SELECT record_rowid, COUNT(*) AS shared_characters
           FROM knowledge_character_index
           WHERE character IN (${characters.map(() => "?").join(", ")})
           GROUP BY record_rowid
           ORDER BY shared_characters DESC, record_rowid
           LIMIT 20
         ) AS candidate ON candidate.record_rowid = knowledge_fts.rowid`,
      )
      .all(...characters) as Row[];
    return candidates
      .map((row) => {
        const searchable = `${text(row, "title")} ${text(row, "summary")} ${text(row, "tags")}`;
        return {
          row,
          score: fuzzyTerms.filter((term) =>
            hasCloseSubstring(searchable, term),
          ).length,
          rowId: Number(row.rowid),
        };
      })
      .filter((candidate) => candidate.score > 0)
      .sort(
        (left, right) => right.score - left.score || left.rowId - right.rowId,
      )
      .slice(0, limit)
      .map((candidate) => candidate.row);
  }
}
